shared/engine/request.js

```diff
@@ -0,0 +1,99 @@
+// @flow
+// Incoming and outgoing requests that are in a session
+import type {MethodKey, ResponseType} from './index'
+import type {invokeType} from './platform-specific'
+
+type SimpleWaiting = (waiting: boolean) => void
+
+// Base class. Handles method and parameters. Waiting callback
+class Request {
+  // RPC call name
+  _method: MethodKey;
+  // RPC parameters
+  _param: Object;
+  // Let others know our waiting state
+  _waitingHandler: (waiting: boolean) => void;
+  // If we're waiting for a response
+  _waiting: boolean = false;
+
+  constructor (method: MethodKey, param: ?Object, waitingHandler: SimpleWaiting) {
+    this._method = method
+    this._param = param || {}
+    this._waitingHandler = waitingHandler
+  }
+
+  get param () { return this._param }
+  get method () { return this._method }
+
+  updateWaiting (waiting: boolean): void {
+    this._waiting = waiting
+    this._waitingHandler(waiting)
+  }
+}
+
+class IncomingRequest extends Request {
+  // Callback in the incomingCallMap
+  _handler: (param: ?Object, request: ResponseType) => void;
+  _response: ?ResponseType;
+
+  constructor (method: MethodKey, param: ?Object, response: ?ResponseType, waitingHandler: SimpleWaiting, handler: Function) {
+    super(method, param, waitingHandler)
+
+    this._handler = handler
+    this._response = response
+  }
+
+  _cleanup () {
+    this.updateWaiting(true) // We just responded to the server so now we're waiting
+  }
+
+  cancel (...args: Array<any>) {
+    this._response && this._response.cancel(...args)
+    this._cleanup()
```

Comment:
so if the server tells us to cancel, and we update waiting to true. We are no longer waiting on the server, right?

I feel like this might be related with my earlier question on waiting